src/io/MemoryProfiler.cpp

```diff
@@ -67,72 +63,46 @@ void MemoryProfiler::doOutput(const std::string& myString) {
 
 unsigned long long MemoryProfiler::getCachedSize() {
 	unsigned long long cached_size = 0;
-	const size_t MAXLEN = 1024;
-	FILE *fp;
-	char buf[MAXLEN];
-	fp = fopen("/proc/meminfo", "r");
-	while (fgets(buf, MAXLEN, fp)) {
-		char *p1 = strstr(buf, "Cached:");
-		if (p1 != NULL) {
-			int colon = ':';
-			p1 = strchr(buf, colon) + 1;
-			cached_size = strtoull(p1, NULL, 10);
+	std::ifstream meminfo_file("/proc/meminfo");
+	std::string token;
+	while(meminfo_file >> token) {
+		if (token == "Cached:") {
+			meminfo_file >> cached_size;
 			break;
 		}
 	}
-	fclose(fp);
 	return cached_size;
 }
 
-int MemoryProfiler::parseLine(char* line) {
-	// This assumes that a digit will be found and the line ends in " Kb".
-	int i = strlen(line);
-	const char* p = line;
-	while (*p < '0' || *p > '9')
-		p++;
-	line[i - 3] = '\0';
-	i = atoi(p);
-	return i;
-}
-
 int MemoryProfiler::countHugePages() {
-	FILE* file = fopen("/proc/self/numa_maps", "r");
-	if (!file) {
-		return 0;
-	}
+	std::ifstream numa_maps_file("/proc/self/numa_maps");
+
 	int hugepagecount = 0;
-	char line[1024];
-	while (fgets(line, 1024, file) != NULL) {
+	std::string line;
+	while(std::getline(numa_maps_file, line)) {
 		//if (/huge.*dirty=(\d+)/) {
-		std::string linestring(line);
-		if (linestring.find("huge") != std::string::npos) {
-			auto pos = linestring.find("dirty");
-			if (pos != std::string::npos) {
-				linestring = linestring.substr(pos);
-				auto eqpos = linestring.find("=");
-				auto spacepos = linestring.find(" ");
-				linestring = linestring.substr(eqpos + 1, spacepos - eqpos - 1);
-				hugepagecount += std::stoi(linestring);
+		if (line.find("huge") != std::string::npos) {
+			const std::regex re("dirty=(\\d+) ");
```

Comment:
Should we add a whitespace before "dirty"?

Reply:
According to the documentation (https://linux.die.net/man/5/numa_maps) we are safe without the whitespace and have no ambiguities here.